sqlglot/optimizer/qualify_columns.py

```diff
@@ -380,18 +410,105 @@ def _qualify_columns(scope: Scope, resolver: Resolver) -> None:
                     column.set("table", column_table)
 
 
+def _expand_struct_stars(
+    expression: exp.Dot,
+    next_alias_name: t.Callable[[], str],
+) -> t.List[exp.Alias]:
+    """[BigQuery] Expand/Flatten foo.bar.* where bar is a struct column"""
+
+    dot_column = t.cast(exp.Column, expression.find(exp.Column))
+    if not dot_column.is_type(exp.DataType.Type.STRUCT):
```

Comment:
where are types being annotated so that this works?

Reply:
The scope is annotated in advance at the entry point of `_expand_stars` in case (1) the dialect is BigQuery and (2) we have `exp.Dot` star expressions i.e struct expansions